pkg/controller/modeler/modeler.go

```diff
@@ -0,0 +1,227 @@
+		Pluck("id", &ids).
+		Error; err != nil {
+		return fmt.Errorf("failed to fetch ids: %w", err)
+	}
+	logger.Debugw("building models", "count", len(ids))
+
+	// Process all models.
+	var merr *multierror.Error
+	for _, id := range ids {
+		if err := c.rebuildModel(ctx, id); err != nil {
+			merr = multierror.Append(merr, fmt.Errorf("failed to update realm %d: %w", id, err))
+		}
+	}
+
+	return merr.ErrorOrNil()
+}
+
+// rebuildModel rebuilds the model for a single model.
+func (c *Controller) rebuildModel(ctx context.Context, id uint64) error {
+	logger := c.logger.Named("rebuildModel").With("id", id)
+	db := c.db.RawDB()
+
+	// Get 21 days of historical data for the realm.
+	var ys []float64
+	if err := db.
+		Model(&database.RealmStats{}).
+		Where("realm_id = ?", id).
+		Order("date DESC").
+		Limit(21).
+		Pluck("codes_issued", &ys).
+		Error; err != nil {
+		return fmt.Errorf("failed to get stats: %w", err)
+	}
+
+	// Require some reasonable number of days of history before attempting to
+	// build a model.
+	if l := len(ys); l < 14 {
+		logger.Warnw("skipping, not enough data", "points", l)
+		return nil
+	}
+
+	// Exclude the most recent record. Depending on timezones, the "day" might not
+	// be over at 00:00 UTC, and we don't want to generate a partial model.
+	ys = ys[:len(ys)-1]
+
+	// Reverse the list - it came in reversed because we sorted by date DESC, but
+	// the model expects the date to be in ascending order.
+	for i, j := 0, len(ys)-1; i < j; i, j = i+1, j-1 {
+		ys[i], ys[j] = ys[j], ys[i]
+	}
+
+	// Build the list of Xs and Ys.
+	xs := make([]float64, len(ys))
+	for i := range ys {
+		xs[i] = float64(i)
+	}
+
+	// This is probably overkill, but it enables us to pick a different curve in
+	// the future, if we want.
+	degree := 2
+	alpha := vandermonde(xs, degree)
+	beta := mat64.NewDense(len(ys), 1, ys)
+	gamma := mat64.NewDense(degree+1, 1, nil)
+	qr := new(mat64.QR)
+	qr.Factorize(alpha)
+	if err := gamma.SolveQR(qr, false, beta); err != nil {
+		return fmt.Errorf("failed to solve QR: %w", err)
+	}
+
+	// Build the curve function.
+	m := gamma.RawMatrix()
+	curve := func(x float64) float64 {
+		var result float64
+		for i := len(m.Data) - 1; i >= 0; i-- {
+			result += m.Data[i] * math.Pow(x, float64(i))
+		}
+		return result
+	}
+
+	// Calculate the predicted next value.
+	next := uint(math.Ceil(curve(float64(len(ys)))))
+
+	// This should really never happen - it means there's been a very sharp
+	// decline in the number of codes issued. In that case, we want to revert
+	// back to the default minimum.
+	if next < 10 {
+		logger.Warnw("next is less than 10, using 10", "next", next)
+		next = 10
+	}
+
+	// Save the new value back, bypassing any validation.
+	if err := db.
+		Model(&database.Realm{}).
+		Where("id = ?", id).
+		UpdateColumn("abuse_prevention_limit", uint(next)).
+		Error; err != nil {
+		return fmt.Errorf("failed to save model: %w", err)
+	}
+
+	// Update the limiter to use the new value.
```

Comment:
nit, should this be done in a function called something other than rebuildModel?

Reply:
Why?

Reply:
`rebuildModel` sounds to me data processing only, and whether to update limiter or not is the decision of the caller. It's not important though, just a nit

Reply:
Ah got it